### DISCUSSION TITLE
指令编号错误

### DISCUSSION THREAD

**Discussion by junshen1** [arcsysu/SYsU-lang#45] [Q&A] [answered]
while判断指令对应的编号错误：
<stdin>:46:3: error: instruction expected to be numbered '%3'
  %2 = icmp slt i32 %tmp3, %tmp2
出现这个错误的原因是什么呢？应该怎么处理呢？

**Comment by junshen1** [comment]
好像不只出现在while中，有些涉及到跳转的部分也会出现这个问题

**Comment by Yorkking** [comment, marked as answer]
LLVM IR 要求虚拟寄存器的编号必须递增命名，并且匿名基本块也会占用一个编号。你可以检查一下你这条 ir 之前是否存在创建过匿名基本块的情况：比如一个基本块已经有终止指令了（br, ret 之类），后面的指令如果没有安排在一个显式声明的基本块中，就好自动创建一个匿名基本块，也占用一个编号。

**Reply by Yorkking** [reply]
> LLVM IR 要求虚拟寄存器的编号必须递增命名，并且匿名基本块也会占用一个编号。你可以检查一下你这条 ir 之前是否存在创建过匿名基本块的情况：比如一个基本块已经有终止指令了（br, ret 之类），后面的指令如果没有安排在一个显式声明的基本块中，就好自动创建一个匿名基本块，也占用一个编号。
参考：https://llvm.org/docs/LangRef.html
![image](https://user-images.githubusercontent.com/34130042/169638999-28d6f83c-63f5-412b-ad05-36cdbd95b526.png)
